.github/actions/setup-spack-stack/action.yaml

```diff
@@ -44,32 +44,108 @@ runs:
             rm GPG-PUB-KEY-INTEL-SW-PRODUCTS.PUB
             echo "deb https://apt.repos.intel.com/oneapi all main" | sudo tee /etc/apt/sources.list.d/oneAPI.list
             sudo apt-get update
-            sudo apt-get install intel-oneapi-dev-utilities intel-oneapi-mpi-devel intel-oneapi-openmp intel-oneapi-compiler-fortran intel-oneapi-compiler-dpcpp-cpp-and-cpp-classic
+            sudo apt-get install intel-oneapi-dev-utilities intel-oneapi-openmp intel-oneapi-compiler-fortran intel-oneapi-compiler-dpcpp-cpp-and-cpp-classic intel-oneapi-mpi-devel
             echo "source /opt/intel/oneapi/setvars.sh" >> ~/.bash_profile
         fi
       elif [ "$RUNNER_OS" == "macOS" ]; then
-        if [[ "${{ inputs.compiler }}" == "clang"* ]]; then
-            brew install llvm
-        fi
         echo ""
       fi
 
       # Install Python poetry to avoid install errors in spack
       python3 -m pip install poetry
 
+  - name: cache-mpi
+    id: cache-mpi
+    uses: actions/cache@v2
+    with:
+      path: ~/mpi
+      key: mpi-${{ inputs.mpi }}-${{ inputs.compiler }}.${{ runner.os }}1
+
+  # MPI takes a long time to build. Built it externally and cache it.
+  - name: install-mpi
+    shell: bash
+    if: steps.cache-mpi.outputs.cache-hit != 'true'
+    run: |
+        echo "" >> ~/.bash_profile
+
+        if [[ "${{ inputs.compiler }}" == "gcc@9"* ]]; then
+          export CC=gcc-9
+          export FC=gfortran-9
+          export CXX=g++-9
+        elif [[ "${{ inputs.compiler }}" == "gcc@10"* ]]; then
+          export CC=gcc-10
+          export FC=gfortran-10
+          export CXX=g++-10
+          export FFLAGS="-fallow-argument-mismatch"
+        elif [[ "${{ inputs.compiler }}" == *"clang"* ]]; then
+          export CC=clang
+          export CXX=clang++
+          export FC=gfortran-10
```

Comment:
I just chose to use gcc-10 with Clang.

Reply:
How is that `install-mpi` step switching between native `apple-clang` and llvm `clang`?

Reply:
Also, on macOS, you could simply use `brew install mpich` or `brew install openmp`, or does that not work?

Reply:
That's a good point. That needs to be corrected.

Reply:
You could install using brew, and same for Ubuntu with APT. But will it change compiler? Will compiler will Brew use out of gcc-9, gcc-10, Clang, and Apple-Clang? FIgure it's just better to build. It's cached, so it's built once and you forget about it.